### PR TITLE
[10.x] Adds testing helpers for Precognition

### PR DESCRIPTION
**Pull request by peterfox**
This should help anyone wanting to write a test for an endpoint using Precognition.

```php
$this->usingPrecognition()
    ->get('/')
    ->assertPrecognitionSuccess();
```

### CODE REVIEW

File: src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php [hunk 1]
```diff
@@ -309,6 +309,16 @@ public function from(string $url)
...
+    /**
+     * Set the precognition header
+     *
+     * @return $this
+     */
+    public function usingPrecognition()
+    {
+        return $this->withHeader('Precognition', 'true');
+    }
+
```
Comment (timacdonald) [inline src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php:321]: Casual suggestion: For testing helpers like this we generally use the `withX`. So we could use:

```suggestion
    /**
     * Set the precognition header
     *
     * @return $this
     */
    public function withPrecognition()
    {
        return $this->withHeader('Precognition', 'true');
    }

```

File: src/Illuminate/Testing/TestResponse.php [hunk 2]
```diff
@@ -1417,6 +1417,25 @@ public function assertSessionMissing($key)
...
+    public function assertPrecognitionSuccess()
+    {
+        PHPUnit::assertTrue(
+            $this->headers->has('Precognition-Success'), 'Precognition-Success Header not present on response.'
+        );
+
+        PHPUnit::assertSame(
+            'True', $this->headers->get('Precognition-Success'),
+            'Precognition-Success Header was found, but the value is not `True`.'
+        );
```
Comment (timacdonald) [inline src/Illuminate/Testing/TestResponse.php:1434]:
```suggestion
        $this->assertStatus(204);
        
        PHPUnit::assertTrue(
            $this->headers->has('Precognition-Success'), 'Precognition-Success Header not present on response.'
        );

        PHPUnit::assertSame(
            'true', $this->headers->get('Precognition-Success'),
            'Precognition-Success Header was found, but the value is not `true`.'
        );
```

File: tests/Foundation/Testing/Concerns/MakesHttpRequestsTest.php [hunk 3]
```diff
@@ -186,6 +186,13 @@ public function testFollowingRedirectsTerminatesInExpectedOrder()
 
         $this->assertEquals(['from', 'to'], $callOrder);
     }
+
+    public function testUsingPrecognition()
+    {
+        $this->usingPrecognition();
+
+        $this->assertSame('true', $this->defaultHeaders['Precognition']);
```
Comment (timacdonald) [inline tests/Foundation/Testing/Concerns/MakesHttpRequestsTest.php:194]:
```suggestion
        $this->withPrecognition();

        $this->assertSame('true', $this->defaultHeaders['Precognition']);
```

File: tests/Testing/TestResponseTest.php [hunk 4]
```diff
@@ -1035,6 +1035,32 @@ public function testAssertHeaderMissing()
...
+
+        $response = TestResponse::fromBaseResponse($baseResponse);
+
+        $response->assertPrecognitionSuccess();
+    }
+
+    public function testAssertPrecognitionSuccessfulWithIncorrectValue()
+    {
+        $this->expectException(AssertionFailedError::class);
+        $this->expectExceptionMessage('Precognition-Success Header was found, but the value is not `True`.');
```
Comment (timacdonald) [inline tests/Testing/TestResponseTest.php:1053]:
```suggestion
        $this->expectExceptionMessage('Precognition-Success Header was found, but the value is not `true`.');
```

**Comment by timacdonald** [comment]
This is awesome, Peter. I've put a few suggestions in. Tomorrow I'll come in and do a good review to make sure it's good to go.

Tried to today, but my Git foo failed me...so unfortunately I just added a bunch of GitHub comments instead of just making the changes - sorry!

Let's leave this in review until I've had a chance to go over it in detail.

Thanks again. Very excited to see this.

File: src/Illuminate/Testing/TestResponse.php (hunk 2 above)
Comment (peterfox) [reply src/Illuminate/Testing/TestResponse.php:1434]: While yes, a Precognition response is, by default, a 204 status. This assertion should only look at the header. One of the reasons I was playing around with this was I wanted to see if I could customise the response to provide additional data in the event of a successful request. If someone wants to assert the status code, it should be done as part of the test, and keep this checking for the header.

**Comment by peterfox** [comment]
@timacdonald thanks for reviewing! I've made the changes bar one that I left a comment on about asserting the response is a 204 status code.

File: src/Illuminate/Testing/TestResponse.php (hunk 2 above)
Comment (timacdonald) [reply src/Illuminate/Testing/TestResponse.php]: Hey Peter,

My concern with not asserting the status is that the precognition libraries check specifically for both a `204` status and `Precognition-Success` header - so a "Precognition Success" involves both of these in tandem.

https://github.com/laravel/precognition/blob/b4f47f7d933674771ef38d83fd2eca5c6f77620d/packages/core/src/client.ts#L18

I feel we need to replicate this check.

Regarding the functionality you are playing around with, you could return any response from a precognition request and respond to it in your application using the `onSuccess` configuration option. You don't need to include the `Precognition-Success` header.

```js
const form = useForm('post', '/users', {
    name: '',
    email: '',
}, {
    onSuccess(response) {
        // Will be invoked when the response is > 200 and < 300 
    },
});
```

Comment (peterfox) [reply src/Illuminate/Testing/TestResponse.php]: I meant making custom precognition Middleware/Controllers on the backend 😅 .

I've been working with customising the Middleware so specific routes can return small objects in response to a successful precognition response. At that point, I would change the response from a 204 to a 200.

Comment (peterfox) [reply src/Illuminate/Testing/TestResponse.php]: Maybe a good compromise would be to have an override boolean for the status? e.g.

```php
    public function assertPrecognitionSuccess(bool $assertStatus = true)
    {
         if ($assertStatus)
         {
             $this->assertNoContent();
         }
    }
```

Comment (timacdonald) [reply src/Illuminate/Testing/TestResponse.php]: I totally see what you are trying to do but I would say that the the `204` and `Preocognition-Success` header is in place to indicate that the response got through all the user specific code  / middleware etc. and ended up back with the framework, i.e., no user returned response was intercepted.

What you are doing is totally valid and really cool to see people experiementing. I imagine it is something like...

```php
class MyMiddleware
{
    public function handle($request, $next)
    {
        if ($request->isPrecognitive()) {
            return response(['custom' => 'data']);
        }

        return $next($response);
    }
}
```

In this case I would argue that this response should not have a `"Precognition-Success"` header and is just a response returned for Precognitive requests.

I probably need to document the "Precognition Specification" to make this all clearer.

What you are doing is awesome, I just don't think you need to add the "Precognition-Success" header to the response.

I am totally keen for you to push back on all and any of this. I don't want to be shutting this down. Here to learn more if you feel there is something I am missing ❤️

Comment (peterfox) [reply src/Illuminate/Testing/TestResponse.php]: That all seems a fair enough explanation 🙂 I'll add the status assertion in.

**Review by timacdonald — APPROVED** [review]
Thanks, Peter!

**Pull request merged**